plugins/dialog/plugin.js

```diff
@@ -1488,6 +1491,62 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
 			}
 
 			this.fire( 'state', state );
+		},
+
+		/**
+		 * @inheritdoc CKEDITOR.dialog.definition#getModel
+		 */
+		getModel: function( editor ) {
+			// Prioritize custom model definition.
+			if ( this.definition.getModel ) {
+				return this.definition.getModel( editor );
+			}
+
+			return this._.model || null;
+		},
+
+		/**
+		 * @since 4.12.0
+		 * @param {CKEDITOR.dom.element/CKEDITOR.plugins.widget} newModel Model to be set.
+		 */
+		setModel: function( newModel ) {
```

Comment:
Agree, it should have the same signature as `dialog.getModel` return type.

Reply:
Actually I'm more concerned about the fact that we don't allow custom `setModel`. If someone creates `getModel` that does not use `this._.model`, this setter would be useless and can introduce unexpected behaviour. Due to that we should introduce the same handling logic as for `getModel`:

```javascript
// Prioritize custom model definition.
if ( this.definition.setModel ) {
	return this.definition.setModel( newModel );
}

return this._.model = newModel
```

Inconsistency in signatures is another, separate issue.

Reply:
I confirm the issue here, but I don't see how your proposition fixes it :thinking: Despite the ability to overwrite `dialog.setModel` by a dialog definition, the cached model still won't be used by `getModel` function
https://github.com/ckeditor/ckeditor-dev/blob/81affc2e0d964b962e554ea93bdaa1b9760cc5be/plugins/dialog/plugin.js#L1499-L1506
so the function is still useless when dialog definition has defined `definition.getModel` function.

I would be happy to get rid of the `setModel` and restore `openDialog` signature, but I couldn't get around the issue with `widget` plugin. 

I think that we could make `setModel` method private, and change `openDialog` function so it can only force the dialog model for the current opening. The API would make more sense because now you can force dialog with `openDialog` method, but without the passed model, the function still uses `defnition.getModel` under the hood.

We could eventually modify widget plugin so it only forces dialog when there is no `definition.getModel` function, although I think that widget dialog should actually always return widget, so the API is more predictable.

I will modify the code with the above solution, please give your opinion if you also think that it should fix the issue with `setModel` method.

Reply:
Probably we're talking about two different issues here.

My concern is: if someone overrides `getModel` to fetch model from elsewhere (not `this._.model`), then not overridden `setModel` does not make sense (as changes introduced by invoking it won't be reflected in `getModel`). And vice versa: if someone overrides `setModel`, then not overridden `getModel` won't fetch new model. If we allow overriding one of them, we de facto must allow overriding both, to keep them in sync.

So something like:

```javascript
getModel: function() {
	if ( definition.getModel ) {
		return definition.getModel();
	}

	return this._.model;
},

setModel: function( value ) {
	if ( definition.setModel ) {
		return definition.setModel( value );
	}

	return this._.model = value;
}
```

IMO there are two fixes for this issue:
* make them both overridable,
* make them both not overridable.

There is also third way: drop setter altogether and leave only `getModel`. All logic connected with getting current model can be enclosed in `getModel`. Note also that currently we use forced model only in `widget` plugin, where it also can be replaced by getting appropriate widget with `getModel`.

Reply:
It seems like a widget is not available anywhere when inserting by dialog (first insertion, not editing) until inserted into the editor. It may be a bit complicated to use only `getModel` for this purpose, and extended `openDialog` adds some small API improvements, although more for edge cases (like our widget). 

Keeping `setModel` private should cover the issue with overwriting getter and setter, also, `dialog.getModel` is not supposed to be shadowed by users (`definition.getModel` should be used instead).